plugins/org.eclipse.xtext.java/src/org/eclipse/xtext/java/resource/JavaResource.xtend

```diff
@@ -0,0 +1,103 @@
+package org.eclipse.xtext.java.resource
+
+import com.google.common.io.CharStreams
+import com.google.inject.Inject
+import com.google.inject.Provider
+import java.io.IOException
+import java.io.InputStream
+import java.io.InputStreamReader
+import java.util.Map
+import org.eclipse.emf.common.util.URI
+import org.eclipse.emf.ecore.InternalEObject
+import org.eclipse.emf.ecore.resource.Resource
+import org.eclipse.emf.ecore.resource.impl.ResourceImpl
+import org.eclipse.emf.ecore.util.EcoreUtil
+import org.eclipse.jdt.internal.compiler.batch.CompilationUnit
+import org.eclipse.xtext.common.types.JvmTypeReference
+import org.eclipse.xtext.common.types.access.IJavaSchemeUriResolver
+import org.eclipse.xtext.common.types.access.impl.AbstractJvmTypeProvider
+import org.eclipse.xtext.common.types.access.impl.IndexedJvmTypeAccess
+import org.eclipse.xtext.common.types.access.impl.URIHelperConstants
+import org.eclipse.xtext.parser.IEncodingProvider
+import org.eclipse.xtext.resource.CompilerPhases
+
+class JavaResource extends ResourceImpl implements IJavaSchemeUriResolver {
+	
+	static class Factory implements Resource.Factory {
+		
+		@Inject Provider<JavaResource> resourceProvider
+		
+		override createResource(URI uri) {
+			resourceProvider.get => [
+				URI = uri
+			]
+		}
+	}
+	
+	@Inject IEncodingProvider encodingProvider
+	@Inject CompilerPhases compilerPhases
+	@Inject JavaDerivedStateComputer derivedStateComputer
+	
+	CompilationUnit compilationUnit
+	
+	override protected doLoad(InputStream inputStream, Map<?, ?> options) throws IOException {
+		// add a dummy eobject as root
+		val encoding = encodingProvider.getEncoding(getURI)
```

Comment:
The encoding may have been passed as an option to load(..), see XtextResource.setEncodingFromOptions(Map<?, ?>)